lxd/metadata/configuration.yaml

```diff
@@ -0,0 +1,1103 @@
+            type: string
+        - volatile.<name>.last_state.vf.spoofcheck:
+            longdesc: The original spoof check setting used when moving a VF into an instance.
+            shortdesc: SR-IOV virtual function original spoof check setting
+            type: string
+        - volatile.<name>.last_state.vf.vlan:
+            longdesc: The original VLAN used when moving a VF into an instance.
+            shortdesc: SR-IOV virtual function original VLAN
+            type: string
+        - volatile.apply_nvram:
+            longdesc: ""
+            shortdesc: Whether to regenerate VM NVRAM the next time the instance starts
+            type: string
+        - volatile.apply_template:
+            longdesc: The template with the given name is triggered upon next startup.
+            shortdesc: Template hook
+            type: string
+        - volatile.base_image:
+            longdesc: The hash of the image that the instance was created from (empty if the instance was not created from an image).
+            shortdesc: Hash of the base image
+            type: string
+        - volatile.cloud_init.instance-id:
+            longdesc: ""
+            shortdesc: '`instance-id` (UUID) exposed to `cloud-init`'
+            type: string
+        - volatile.evacuate.origin:
+            longdesc: The cluster member that the instance lived on before evacuation.
+            shortdesc: The origin of the evacuated instance
+            type: string
+        - volatile.idmap.base:
+            longdesc: ""
+            shortdesc: The first ID in the instance's primary idmap range
+            type: integer
+        - volatile.idmap.current:
+            longdesc: ""
+            shortdesc: The idmap currently in use by the instance
+            type: string
+        - volatile.idmap.next:
+            longdesc: ""
+            shortdesc: The idmap to use the next time the instance starts
+            type: string
+        - volatile.last_state.idmap:
+            longdesc: ""
+            shortdesc: Serialized instance UID/GID map
+            type: string
+        - volatile.last_state.power:
+            longdesc: ""
+            shortdesc: Instance state as of last host shutdown
+            type: string
+        - volatile.uuid:
+            longdesc: The instance UUID is globally unique across all servers and projects.
+            shortdesc: Instance UUID
+            type: string
+        - volatile.uuid.generation:
+            longdesc: |-
+                The instance generation UUID changes whenever the instance's place in time moves backwards.
+                It is globally unique across all servers and projects.
+            shortdesc: Instance generation UUID
+            type: string
+        - volatile.vsock_id:
+            longdesc: ""
+            shortdesc: Instance `vsock ID` used as of last start
+            type: string
+    server-acme:
+        - acme.agree_tos:
+            default: '`false`'
```

Comment:
The backticks turn them into code in Markdown. And the quotes are needed to make it valid YAML afair.

Reply:
Different question: Why is this file checked in? Shouldn't it be generated dynamically?

Reply:
> Different question: Why is this file checked in? Shouldn't it be generated dynamically?

I understand @gabrielmougard wanted it to be generated by the developer and committed, like we do with the API docs, and i18n translations.

Reply:
> The generated `.txt` document (used for the site documentation), is derived from the YAML document (the marshaled YAML is the 'source of truth') in the code.

What tool does this conversion? It seems to me that this tool should be the one converting it to markdown, rather than being double encoded in the yaml file that is also served out through the API for non-documentation consumers (e.g. UI).

Reply:
There is a semantic difference, so I think it should be in the code comment.
Think about a default value - "no" could mean that there is no default value, while "`no`" means the default value is `no`.

Reply:
The "no" was just one example. It could also be "auto" - which can mean either the value `auto` or that the value is automatically determined.

Forget about what I said about Markdown. It's really a differentiation between a specific value and a description.

Reply:
We also have things like ``default: `10` (maximum)``, for example, which has both a value and a description.

Reply:
@ru-fu ah OK I see the challenge.

Because this is a metadata API rather than purely for documentation generation, it sounds to me like we need to come up with a slightly different schema that includes additional fields that can enable us to describe programmatically what the default values are and what the valid values/ranges are for a field (at a basic level).

So to check I understand the things we're trying to represent here, they are:

- What type of default value it has (potentially a list of defined values that indicate whether there is a default at all or if the value is automatically determined).
- The static default value (if there is one) - which would be its literal value. A question I have here is whether we should use YAML's built in data types to represent the value (i.e boolean for true/false, int for numbers, strings for text etc) or whether we should always present as a string (because that is what the API will accept for configuration items).
- One or more optional fields that allow describing the basic validation of the field (i.e list of valid literal values or ranges of numbers).

This sounds a bit like the schema we designed for ACL rules, but more general:

https://documentation.ubuntu.com/lxd/en/latest/howto/network_acls/#rule-properties

Then the tool that converts the yaml file to the documentation text representation would have the knowledge to take the metadata and convert it to the correct representation for the documentation. And the UI would be able to consume the YAML via the API and generate the form fields (with basic validation) from it.

Reply:
I guess that's the discussion with the UI team that needs to happen.

From a documentation perspective, this might be overshooting it a bit - we just need a series of fields/keys and their content, similar to the tables we had before (and we don't only have the "default" field, but several others - the ones that are currently supported are here: https://github.com/canonical/lxd-sphinx-extensions/blob/9a52e049fd2de70e0451cc725ba7ed2669afbd7c/config-options/__init__.py#L27 ). But as long as the tool generated suitable doc blocks, the input can be more complicated if needed.

Reply:
@gabrielmougard with the above discussion in mind please can you have a think about possible approaches here and update the spec post so we can have something to discuss with the UI team. Thanks

Reply:
Hi everyone,

I'm wondering whether there is a way for us to expand the schema also for storage pools?

Also, would it make sense to have a field `managed` (or something like this) in the schema indicating that if the field is not set, it will be managed by the LXD itself. For some settings this cannot be represented with the default value. For example, IP address does not have a default value, but the value will be set by the LXD (if possible of course).

Example use case:

Having this for storage (potentially for all storage drivers) would allow external applications to know ahead of time which settings will be set by LXD. To be more specific, in `terraform-lxd-provider` there is currently no way to detect which settings are managed by the LXD, causing inconsistencies between Terraform runs (see also terraform-lxd/terraform-provider-lxd/issues/328). Or in other words, the user is required to always provide ceratin settings, such as pool's source and size, otherwise Terraform will try to delete those fields when the configuration is reapplied.

Reply:
Seems like this is already in the schema - I've somehow missed the link provided by @ru-fu: https://github.com/canonical/lxd-sphinx-extensions/blob/9a52e049fd2de70e0451cc725ba7ed2669afbd7c/config-options/__init__.py#L27

Reply:
Yes, from a documentation perspective, we want to use this tool to generate all configuration option tables, including the ones for the storage pools.
Here's the list of which are still missing: https://warthogs.atlassian.net/browse/LXD-430

Reply:
Thanks for the link, can I help with some subtasks?

Reply:
We should probably wait for @gabrielmougard's updated spec proposal on schema before we start updating the code itself as it might mean additional rework later. We're also due to have a meeting with the UX team to discuss their needs.

Reply:
I would only say "growing" WRT to @MusicDin , but the original requirement has always been for the UI to consume the YAML file via the API programmatically, and so in that sense the original requirements remain unchanged.

I don't think we need anything "private", but what we need is to remove any markdown encoding in the YAML response.

In doing this we will encounter the problem that currently the markdown in the `default` field is encoding multiple pieces of information in human readable form, and its that which is problematic here.

Instead we should introduce more fields to the YAML file that allow for encoding the information needed to *both* generate the resulting documentation markdown (in the current format it is) *and* to allow the UX and Terraform providers to be able to parse the information about each field programmatically.

Reply:
I like your approach, however, to me it would seem cleaner to have a `default` field and `defaultDoc`, `defaultDisplay`, or maybe simply `display`. This is just a thought, but this way, the difference between the fields would be clear from their names.

Reply:
I'd like to try and avoid duplicating the default info if we can. Really the markdown values being put in the default field at the moment are expressing things beyond the default value, such as validation rules. So rather than duplicating, we should try and find a way to express the validation info in a separate field.

I believe we could also use the absence of a default field to indicate if there is "no" default value.

Reply:
I agree that we should not duplicate any information.
Remember that you'll need to write these comments for every new config option you add, and you don't want to write the same information in two different formats.